### PR TITLE
Fixed #30758 -- Made RangeFields use multiple hidden inputs for initial data.

### PR DESCRIPTION
**Pull request by nasirhjafri**
https://code.djangoproject.com/ticket/30758

### CODE REVIEW

**Comment by felixxm** [comment]
@nasirhjafri Thanks for this PR, however I don't think that it's a proper approach. IMO we should rather fix passing initial data for range fields. Currently we use `HiddenInput` which ends with 
```html
<input type="hidden"
       name="initial-range_field"
       value="[datetime.datetime(2019, 9, 13, 9, 53, 31, 531040), None]"
...
```
IMO we should change `hidden_widget` in `BaseRangeField` to get
```html
<input type="hidden" name="initial-range_field_0" value="2019-09-13 09:53:31" id="initial-id_range_field_0">
<input type="hidden" name="initial-range_field_1" id="initial-id_range_field_1">
```
e.g.
```diff
diff --git a/django/contrib/postgres/forms/ranges.py b/django/contrib/postgres/forms/ranges.py
index 47a16a1bbd..2d55cbce57 100644
--- a/django/contrib/postgres/forms/ranges.py
+++ b/django/contrib/postgres/forms/ranges.py
@@ -2,7 +2,7 @@ from psycopg2.extras import DateRange, DateTimeTZRange, NumericRange
 
 from django import forms
 from django.core import exceptions
-from django.forms.widgets import MultiWidget
+from django.forms.widgets import MultiWidget, MultipleHiddenInput
 from django.utils.translation import gettext_lazy as _
 
 __all__ = [
@@ -16,6 +16,7 @@ class BaseRangeField(forms.MultiValueField):
         'invalid': _('Enter two valid values.'),
         'bound_ordering': _('The start of the range must not exceed the end of the range.'),
     }
+    hidden_widget = MultipleHiddenInput
 
     def __init__(self, **kwargs):
         if 'widget' not in kwargs:
```

**Comment by nasirhjafri** [comment]
@felixxm That's how it was being done in 1.11.x https://github.com/django/django/blob/stable/1.11.x/django/forms/fields.py#L398

I'll try the solution you provided, thanks.

**Comment by felixxm** [comment]
@nasirhjafri Yes but it's clunky, IMO.

**Comment by nasirhjafri** [comment]
@felixxm Any ideas on how test cases should look like? This solution works but I can't think of a test case.

**Comment by felixxm** [comment]
We should test the way how `RangeFields` widgets are rendered with `initial` data (`tests.postgres_tests.TestWidget`) and how `has_changed()` works for these fields with and without `initial` data (see [similar tests](https://github.com/django/django/blob/ea25bdc2b94466bb1563000bf81628dea4d80612/tests/forms_tests/field_tests/test_multivaluefield.py#L96-L119) for `MultiValueField`).

**Comment by nasirhjafri** [comment]
@felixxm I've pushed the tests, but they are not running on circle ci. Can you please manually run the tests on PR?

File: tests/postgres_tests/test_widgets.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        cls.date_field = DateRangeField(widget=RangeWidget(DateInput))
+        cls.datetime_field = DateTimeRangeField(widget=RangeWidget(DateTimeInput))
+        super().setUpClass()
+
+    def test_date_range_has_changed_no_initial(self):
+        self.assertTrue(self.date_field.has_changed(None, ['2010-01-01']))
+
+    def test_date_range_has_changed_empty_initial(self):
+        self.assertTrue(self.date_field.has_changed(['2010-01-01', ''],
+                                                    ['2010-01-01', '2020-12-12']))
```
Comment (felixxm) [inline tests/postgres_tests/test_widgets.py]: Please use hanging indentation per our coding style (in all tests).

**Comment by felixxm** [comment]
buildbot

**Comment by nasirhjafri** [comment]
buildbot

**Comment by nasirhjafri** [comment]
@felixxm This one is failing without the fix and working with the fix.
https://github.com/django/django/blob/3b7febf5edac974287f99d5de3a11fe1874edd3d/tests/postgres_tests/test_ranges.py#L802-L813

I've created two separate commits. Please review.

**Comment by felixxm** [comment]
@nasirhjafri Thanks! I simplified tests, moved them between test classes and added some extra tests. I also notice that with `MultipleHiddenInput` it doesn't detect changes as expected, so I decided to add a new `HiddenRangeWidget` which works fine, IMO. Can you take a look?

**Comment by felixxm** [comment]
buildbot

**Comment by nasirhjafri** [comment]
@felixxm looks good to me, working fine with the new `HiddenRangeWidget`.

**Comment by felixxm** [comment]
Great, I squashed commits and rebased from master.

**Comment by felixxm** [comment]
Merged in faf4b988fe75dd4045bc5c62496cc4f2e0db8c4d and 733dbb21c734920fceb7c285e7396b15db4182ed.

**Comment by yeppus** [comment]
Hi! Thanks a lot for fixing this!

@felixxm do you know if this patch will be able to make it into 2.2.6?

**Comment by felixxm** [comment]
@yeppus No, it doesn't qualify for a backport to 2.2.x, but it's included in Django 3.0.

**Comment by yeppus** [comment]
@felixxm thanks for letting me know,,, we need the LTS so I guess we need to rework the code around this bug in that case.

**Pull request closed without merging**